src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/AesGcm.Unix.cs

```diff
@@ -0,0 +1,202 @@
+        private SafeEvpCipherCtxHandle _ctxHandle;
+
+        public AesGcm(ReadOnlySpan<byte> key)
+        {
+            if (key == null)
+                throw new ArgumentNullException(nameof(key));
+
+            int keyLength = key.Length * 8;
+            AesAEAD.CheckKeySize(keyLength);
+
+            _ctxHandle = Interop.Crypto.EvpCipherCreatePartial(
+                        GetCipher(keyLength),
+                        keyLength,
+                        0);
+
+            Interop.Crypto.CheckValidOpenSslHandle(_ctxHandle);
+
+            ref byte nullRef = ref MemoryMarshal.GetReference(Span<byte>.Empty);
+            if (!Interop.Crypto.EvpCipherSetKeyAndIV(_ctxHandle, ref MemoryMarshal.GetReference(key), ref nullRef, -1))
+            {
+                throw Interop.Crypto.CreateOpenSslCryptographicException();
+            }
+
+            Debug.Assert(NonceByteSizes.MinSize == 12 && NonceByteSizes.MaxSize == 12);
+            if (!Interop.Crypto.EvpAesGcmSetNonceLength(_ctxHandle, 12))
+            {
+                throw Interop.Crypto.CreateOpenSslCryptographicException();
+            }
+        }
+
+        public void Encrypt(ReadOnlySpan<byte> nonce, ReadOnlySpan<byte> plaintext, Span<byte> ciphertext, Span<byte> tag, ReadOnlySpan<byte> associatedData = default)
+        {
+            AesAEAD.CheckArguments(nonce, plaintext, ciphertext, tag);
+            CheckParameters(nonce.Length, tag.Length);
+
+            ref byte nullRef = ref MemoryMarshal.GetReference(Span<byte>.Empty);
+
+            if (!Interop.Crypto.EvpCipherSetKeyAndIV(_ctxHandle, ref nullRef, ref MemoryMarshal.GetReference(nonce), 1 /* encrypting */))
+            {
+                throw Interop.Crypto.CreateOpenSslCryptographicException();
+            }
+
+            if (associatedData.Length != 0)
+            {
+                if (!Interop.Crypto.EvpCipherUpdate(_ctxHandle, ref nullRef, out _, ref MemoryMarshal.GetReference(associatedData), associatedData.Length))
+                {
+                    throw Interop.Crypto.CreateOpenSslCryptographicException();
+                }
+            }
+
+            int ciphertextBytesWritten;
+            if (plaintext.Length != 0)
+            {
+                if (!Interop.Crypto.EvpCipherUpdate(_ctxHandle, ref MemoryMarshal.GetReference(ciphertext), out ciphertextBytesWritten, ref MemoryMarshal.GetReference(plaintext), plaintext.Length))
+                {
+                    throw Interop.Crypto.CreateOpenSslCryptographicException();
+                }
+
+                if (!Interop.Crypto.EvpCipherFinalEx(_ctxHandle, ref MemoryMarshal.GetReference(ciphertext.Slice(ciphertextBytesWritten)), out int bytesWritten))
+                {
+                    throw Interop.Crypto.CreateOpenSslCryptographicException();
+                }
+
+                ciphertextBytesWritten += bytesWritten;
+            }
+            else
+            {
+                byte nonEmptyArray = 123;
```

Comment:
Do we even need to support 0-length plaintext? What's the case for just producing authenticated data with AES?

Reply:
I think for something like TLS that's "I'm sending an empty keepalive frame"

Reply:
@morganbr mixed feeling about supporting vs non-supporting this scenario - on one hand there is probably no good scenarios where you should do that over i.e. calculating hash on the other hand it might make sense if for whatever reason you got such data over the wire (or the other end expected it) and would like to validate it. Since this is a low-level API I do not think we should not support it